File: src/main/java/dragonbook/inter/stmt/If.java
package dragonbook.inter.stmt;

import dragonbook.inter.expr.Expr;
import dragonbook.symbols.Type;

public class If extends Stmt {
    public final Expr condition;
    public final Stmt then;

    public If(Expr condition, Stmt then) {
        if (condition.type != Type.BOOL) condition.error("boolean required in if");
        this.condition = condition;
        this.then = then;
    }

    @Override
    public String toString() {
        return "if (" + condition + ") " + then;
    }
}

File: src/main/kotlin/dragonbook/inter/stmt/If.kt
package dragonbook.inter.stmt

import dragonbook.error.ParseError
import dragonbook.inter.expr.Expr
import dragonbook.symbols.Type

data class If(val condition: Expr, val then: Stmt) : Stmt {

    init {
        validate()
    }

    private fun validate() {
        if (condition.type != Type.BOOL) throw ParseError("boolean required in if", condition)
    }

    override fun toString(): String = "if ($condition) { $then }"
}